File: lib/features/order/domain/repositories/order_repository.dart
import 'package:dartz/dartz.dart';

import '../../../../core/errors/failures.dart';
import '../../../../core/pagination/pagination.dart';
import '../entities/order_entity.dart';

/// Domain repository contract for Order. Implementations live in the
/// `data/` layer.
abstract class OrderRepository {
  Future<Either<Failure, List<OrderEntity>>> getAll();
  Future<Either<Failure, PaginatedResponse<OrderEntity>>> getAllPaginated(
    PaginationParams params,
  );
  Future<Either<Failure, OrderEntity>> getById(String id);
  Future<Either<Failure, OrderEntity>> add(OrderEntity entity);
  Future<Either<Failure, OrderEntity>> update(OrderEntity entity);
  Future<Either<Failure, Unit>> delete(String id);
  Future<Either<Failure, List<OrderEntity>>> search(String query);
}

File: packages/example_react_app/src/features/order/domain/repositories/order.repository.ts
/// Repository contract for Order.
import type { Either } from '../../../../core/usecase/usecase';
import type { Failure } from '../../../../core/errors/failures';
import type { OrderEntity } from '../entities/order.entity';
import type { PaginationParams, PaginatedResponse } from '../../../../core/pagination/pagination';

export interface OrderRepository {
  getAll(): Promise<Either<Failure, OrderEntity[]>>;
  getById(id: string): Promise<Either<Failure, OrderEntity>>;
  add(entity: OrderEntity): Promise<Either<Failure, OrderEntity>>;
  update(entity: OrderEntity): Promise<Either<Failure, OrderEntity>>;
  delete(id: string): Promise<Either<Failure, void>>;
  search(query: string): Promise<Either<Failure, OrderEntity[]>>;
  getAllPaginated(params: PaginationParams): Promise<Either<Failure, PaginatedResponse<OrderEntity>>>;
}
